src/ray/raylet/scheduling/policy/node_affinity_scheduling_policy.cc

```diff
@@ -0,0 +1,39 @@
+// Copyright 2021 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/raylet/scheduling/policy/node_affinity_scheduling_policy.h"
+
+namespace ray {
+namespace raylet_scheduling_policy {
+
+scheduling::NodeID NodeAffinitySchedulingPolicy::Schedule(
+    const ResourceRequest &resource_request, SchedulingOptions options) {
+  RAY_CHECK(options.scheduling_type == SchedulingType::NODE_AFFINITY);
+
+  scheduling::NodeID target_node_id = scheduling::NodeID(options.node_affinity_node_id);
+  if (nodes_.contains(target_node_id) && is_node_alive_(target_node_id) &&
+      nodes_.at(target_node_id).GetLocalView().IsFeasible(resource_request)) {
```

Comment:
That's a good question. The current assumption is that unavailable but feasible node will become available at some point so we prefer waiting instead of immediately schedule it somewhere else. But I can see your point that the wait can be potentially indefinite.

What's your use case and what's the behavior you want?

Reply:
@wumuzi520 that's a very good point. Do you have a specific case in you mind. Currently we assume unavailability is a temporary thing according our current use cases (shuffle). But if you have a counter example we can reevaluate.

There are few options if it becomes problematic, we need to discuss more on this.
- add another parameter to control the soft behavior, as you suggested
- change the policy on how we decide a resource request is infeasible, for example we can assume resource allocated to task is temporary but to actor is permanent.
- add timeout in node_affility_scheduling policy so if a feasible request has been waiting too long we can reschedule it.

Reply:
I agree we need to handle cases such as actors. Probably we could do this by improving the soft policy to consider load on other nodes vs the desired node, more similar to how data locality is usually handled.

My preference here would be to merge as is and then later change the policy to become load-aware. This will help us to unblock existing applications but also keep the API minimal and maintainable. Adding a timeout or other enumerations besides a bool seems pretty tricky to get right in terms of UX.

Reply:
I don't like treating tasks as temporary and actors as permanent. Because tasks can be arbitrarily long as well. Especially that the node may have tasks from other jobs, which are completely out of the current job's control. 
I'm fine with adding new parameters (control the behavior or timeout). But it's a bit complex to implement.

The simplest approach in my mind is to simply make the decision based on the current available resources. This makes sense because the request is "soft". Because I'm assuming this may not meet your scenarios? 

Also, @stephanie-wang, could you elaborate on "consider load on other nodes vs the desired node"? I don't understand that. Thank you.

Reply:
Offline discussion: we can merge as it is and evolve this api based on new use cases.